### PR TITLE
server, pd-ctl: support multiple stores for evict-leader-scheduler

### LINKED ISSUE #11286: evict-leader-scheduler: support adding multiple store IDs in one request
### Describe your feature request related problem

Adding evict-leader-scheduler for multiple stores today requires one `POST /schedulers` call per store (the first call creates the scheduler, subsequent calls are redirected to the per-scheduler config-update endpoint since the scheduler already exists). When an operator needs to evict leaders from several stores at once (e.g. draining a rack or applying a maintenance window across many nodes), this means N sequential HTTP calls (or N `pd-ctl scheduler add evict-leader-scheduler <id>` invocations).

### Describe the feature you'd like

Let `POST /pd/api/v1/schedulers` accept a `store_ids` array (in addition to the existing single-value `store_id` field) for `evict-leader-scheduler`, so one request can evict leaders from multiple stores at once. Correspondingly, `pd-ctl scheduler add evict-leader-scheduler <store_id> [store_id...]` should accept multiple store IDs.

Compatibility requirements:
- The existing single `store_id` field/behavior is unchanged for old clients.
- `store_ids` is a new, optional field; supplying both `store_id` and `store_ids` in the same request is rejected.

### Describe alternatives you've considered

- Keep issuing one request per store from the client side (status quo) — works, but pushes the looping/error-handling burden onto every caller.
- Extend `grant-leader-scheduler` the same way — out of scope here since the immediate need is for evict-leader-scheduler.

### Teachability, Documentation, Adoption, Migration Strategy

No migration needed: `store_id` keeps working exactly as before, so existing scripts/tools are unaffected. `store_ids` is additive.

### PR DESCRIPTION
**Pull request by bufferflies**
### What problem does this PR solve?

Issue Number: Close #11286

### What is changed and how does it work?

```commit-message
Let POST /schedulers and `pd-ctl scheduler add evict-leader-scheduler`
accept several store IDs in one call via a new `store_ids` array field,
instead of requiring one request per store. The existing single
`store_id` field keeps working unchanged for old clients; sending both
fields at once is rejected.

The scheduler is created with the first store ID and the remaining
ones are added to it through the same per-store config-update path
already used when the scheduler exists, so no changes were needed to
the scheduler's config storage or persistence format.

`pd-ctl` mirrors this: it sends `store_id` whenever exactly one ID is
given (so it still talks to an older pd-server for the common case)
and `store_ids` only when adding multiple stores at once.

The single-store and store_ids-batch config-update paths are unified
into one handler sharing a single lock across the whole
check-then-write sequence, so a multi-store update is atomic with
respect to concurrent config updates on the same scheduler, and a
failure partway through (including a mid-batch pause failure, not
just the final persist) rolls back completely: previously existing
stores keep their prior ranges, and only the stores newly added by
that call are removed.
```

### Check List

Tests

- Unit test
- Integration test

Code changes

- Has HTTP APIs changed (Don't forget to add the declarative for the new API)

### Known limitations

- If a multi-store create's follow-up batch-update fails **and** the
  rollback (`RemoveScheduler`) of the just-created scheduler also
  fails for a reason other than "already gone" (e.g. the underlying
  `ConfigStorage.RemoveSchedulerConfig` call itself errors), the
  response now surfaces both failures instead of silently logging the
  rollback error, but there is no automated regression test for this
  specific double-failure case — it would need a fault-injectable
  `ConfigStorage`, which felt disproportionate for this edge case.
- `RedirectSchedulerUpdateBatch` failures always surface as HTTP 500
  from `CreateScheduler`, whether the underlying cause is a client
  error (a batch store ID that doesn't exist, failing in
  `applyStoreIDs`'s `PauseLeaderTransfer`) or a genuine server-side
  persistence failure (`conf.save()`). This matches the pre-existing
  single-store `RedirectSchedulerUpdate` path, which has the same
  imprecision. Distinguishing the two would require the original HTTP
  status code to survive `pkg/utils/apiutil`'s
  `checkResponse`/`PostJSONIgnoreResp`, which currently discard it —
  shared plumbing used by both the single- and multi-store paths, so
  out of scope for this PR.

### Release note

```release-note
`evict-leader-scheduler` now accepts multiple store IDs in a single request: `POST /schedulers` supports a new `store_ids` array field, and `pd-ctl scheduler add evict-leader-scheduler <store_id> [store_id...]` accepts more than one store ID.
```

## Summary by CodeRabbit

* **New Features**
  * Evict-leader scheduling supports targeting multiple stores in one request or command.
  * Store IDs accept individual or list formats with validation for invalid or conflicting values.
  * Multi-store updates apply consistently while preserving existing single-store behavior and configurations.
  * Grant-leader scheduler commands continue to accept only one store ID.
  * Removing an evict-leader scheduler restores leader-transfer behavior for targeted stores.

* **Bug Fixes**
  * Failed multi-store updates now roll back changes to prevent partial configuration.
  * Rollback failures are reported clearly when manual cleanup may be required.

### CODE REVIEW

File: server/api/scheduler.go [hunk 1]
```diff
@@ -256,6 +294,45 @@ func (h *schedulerHandler) CreateScheduler(w http.ResponseWriter, r *http.Reques
...
+		return nil, errors.New("only one of store_id and store_ids can be set")
+	case hasStoreIDs:
+		arr, ok := rawStoreIDs.([]any)
+		if !ok || len(arr) == 0 {
+			return nil, errors.New("please input a right store id")
+		}
+		storeIDs := make([]float64, 0, len(arr))
+		seen := make(map[float64]struct{}, len(arr))
+		for _, v := range arr {
+			id, ok := v.(float64)
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/api/scheduler.go:334]: Good catch, fixed in `1c8ab1e97`: added `isIntegerStoreID` to reject non-integer, negative, or out-of-uint64-range float values before they reach the `uint64()` conversion, applied to both the `store_id` and `store_ids` branches of `collectEvictLeaderStoreIDs`.

File: tools/pd-ctl/pdctl/command/scheduler_command.go [hunk 2]
```diff
@@ -688,7 +697,7 @@ func newConfigShuffleRegionCommand() *cobra.Command {
 }
 
 func addStoreToSchedulerConfig(cmd *cobra.Command, schedulerName string, args []string) {
-	if len(args) != 1 {
+	if len(args) < 1 {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply tools/pd-ctl/pdctl/command/scheduler_command.go]: Good catch, fixed in `1c8ab1e97`: `addStoreToSchedulerConfig` now rejects more than one argument unless `schedulerName == evictLeaderSchedulerName`, restoring the original single-store requirement for `grant-leader-scheduler config add-store` (and any other scheduler reusing this helper).

File: pkg/schedule/schedulers/evict_leader.go [hunk 3]
```diff
@@ -213,6 +213,24 @@ func (conf *evictLeaderSchedulerConfig) update(id uint64, newRanges []keyutil.Ke
...
+	defer conf.Unlock()
+	for _, id := range ids {
+		conf.StoreIDWithRanges[id] = newRanges
+	}
+	conf.Batch = batch
+	err := conf.save()
+	if err != nil {
+		for _, id := range ids {
+			_, _ = conf.removeStoreLocked(id)
+		}
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply pkg/schedule/schedulers/evict_leader.go]: Good catch, fixed in `1c8ab1e97`, though not by patching `updateBatch` directly — I unified it with the single-store `update` path into one `applyStoreIDs` method that tracks which store IDs were newly added in this call versus already existed, and on failure only removes the newly-added ones while restoring the previously-existing ones' prior ranges (and restoring `Batch`). While doing this I also found and fixed a related gap: a failure partway through the id loop (not just at the final save) hadn't been rolled back completely either — see `TestEvictLeaderUpdateConfigBatchRollbackOnMidLoopPauseFailure`, which fails against the pre-fix code.

File: pkg/schedule/schedulers/evict_leader.go [hunk 4]
```diff
@@ -473,6 +497,96 @@ func (handler *evictLeaderHandler) updateConfig(w http.ResponseWriter, r *http.R
 	handler.rd.JSON(w, http.StatusOK, "The scheduler has been applied to the store.")
 }
 
+// updateConfigBatch is the "store_ids" counterpart of updateConfig: it applies
+// the same ranges and batch size to every listed store in a single call and a
+// single persisted save, instead of requiring one request per store. Unlike
+// updateConfig's single-id path, it does not fall back to a store's existing
+// ranges when ranges are omitted — every listed store gets the same ranges
+// (the whole key space by default), which matches its intended use of adding
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply pkg/schedule/schedulers/evict_leader.go]: Good catch, fixed in `1c8ab1e97` as part of the same unification: when `ranges` is omitted, `applyStoreIDs` now keeps each already-existing store's current ranges untouched and only defaults the whole key space for stores newly added by that call, matching the single-store path's behavior.

File: pkg/schedule/schedulers/evict_leader.go [hunk 5]
```diff
@@ -473,6 +497,96 @@ func (handler *evictLeaderHandler) updateConfig(w http.ResponseWriter, r *http.R
...
+	}
+	ids := make([]uint64, 0, len(rawIDs))
+	seen := make(map[uint64]struct{}, len(rawIDs))
+	for _, v := range rawIDs {
+		idFloat, ok := v.(float64)
+		if !ok {
+			handler.rd.JSON(w, http.StatusBadRequest, "please input a right store id")
+			return
+		}
+		id := uint64(idFloat)
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply pkg/schedule/schedulers/evict_leader.go]: Good catch, fixed in `1c8ab1e97`: `parseUpdateStoreIDs` (the new shared parser for both `store_id` and `store_ids`) now validates via `storeIDFromFloat`, rejecting fractional, negative, or out-of-range values before conversion — same fix as the create-endpoint side of this, in `server/api/scheduler.go`.

File: pkg/schedule/schedulers/evict_leader.go [hunk 6]
```diff
@@ -473,6 +497,96 @@ func (handler *evictLeaderHandler) updateConfig(w http.ResponseWriter, r *http.R
...
+			return
+		}
+		batch = (int)(batchFloat)
+	}
+
+	ranges, ok := (input["ranges"]).([]string)
+	if !ok && input["ranges"] != nil {
+		resumeAll()
+		handler.rd.JSON(w, http.StatusBadRequest, fmt.Sprintf("invalid argument for 'ranges': expected an array of strings, got %T", input["ranges"]))
+		return
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply pkg/schedule/schedulers/evict_leader.go]: Good catch, fixed in `1c8ab1e97`: added `decodeStringSlice`, which explicitly converts a JSON-decoded `[]any` into `[]string` (checking each element's type) instead of relying on a direct `.([]string)` assertion that a JSON array can never satisfy. This applies to the single-store path too now, since both are unified into one handler — so a request with `ranges` set works for both a lone `store_id` and a `store_ids` batch.

File: server/api/scheduler.go [hunk 7]
```diff
@@ -207,6 +207,55 @@ func (h *schedulerHandler) CreateScheduler(w http.ResponseWriter, r *http.Reques
...
+					if rmErr := h.RemoveScheduler(name); rmErr != nil {
+						log.Error("failed to roll back partially created evict-leader-scheduler",
+							zap.String("scheduler-name", name), errs.ZapError(rmErr))
+					}
+				}
+				h.r.JSON(w, http.StatusInternalServerError, err.Error())
+				return
+			}
+			log.Info("update scheduler", zap.String("scheduler-name", name), zap.Any("store-ids", toUpdate))
+		}
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/api/scheduler.go:261]: Confirmed and mostly fixed in `1c8ab1e97`: if `RemoveScheduler` itself fails during rollback (and it isn't just reporting the scheduler is already gone), the response now combines both errors and tells the caller the scheduler may still be partially active, instead of silently logging `rmErr` and returning only the original batch-update error. Not adding the requested regression test for `RemoveSchedulerConfig` itself failing — it needs a fault-injectable `ConfigStorage` for a fairly narrow double-failure edge case, and I judged that disproportionate here; noted in the PR description's Known limitations. Happy to add it if you'd still like that level of coverage.

File: tests/server/api/scheduler_test.go [hunk 8]
```diff
@@ -156,6 +156,140 @@ func (suite *scheduleTestSuite) checkOriginAPI(cluster *tests.TestCluster) {
...
+	// (and later) store, not the initial scheduler-creation save, so this
+	// reproduces "first store created, batch-add the rest fails".
+	re.NoError(failpoint.Enable("github.com/tikv/pd/pkg/schedule/schedulers/persistFail", "return(true)"))
+	input := map[string]any{"name": "evict-leader-scheduler", "store_ids": []int{1, 2}}
+	body, err := json.Marshal(input)
+	re.NoError(err)
+	re.NoError(testutil.CheckPostJSON(tests.TestDialClient, urlPrefix, body,
+		testutil.Status(re, http.StatusInternalServerError)),
+	)
+	re.NoError(failpoint.Disable("github.com/tikv/pd/pkg/schedule/schedulers/persistFail"))
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply tests/server/api/scheduler_test.go]: Good catch, fixed in `1c8ab1e97`: `failpoint.Disable` is now registered via `defer` immediately after `failpoint.Enable`, so it still runs if an assertion in between fails.

File: pkg/schedule/schedulers/evict_leader.go [hunk 9]
```diff
@@ -180,37 +181,69 @@ func (conf *evictLeaderSchedulerConfig) resumeLeaderTransfer(cluster sche.Schedu
...
+				conf.StoreIDWithRanges[id] = []keyutil.KeyRange{keyutil.NewKeyRange("", "")}
+			}
+			continue
+		}
+		if !hasExplicitRanges {
+			// Keep the existing store's current ranges untouched.
+			continue
+		}
+		prevRanges[id] = old
+		conf.StoreIDWithRanges[id] = explicitRanges
```
Comment (lhy1024) [inline pkg/schedule/schedulers/evict_leader.go]: Every store in a `store_ids` + `ranges` update is assigned the same `explicitRanges` backing array (both the newly added and the already-existing branches below). I confirmed this with a unit test: after `POST {"store_ids":[1,2],"ranges":["c","d"]}`, mutating `StoreIDWithRanges[1][0].StartKey` also changes store 2.

Please copy per store, e.g.

```go
conf.StoreIDWithRanges[id] = append([]keyutil.KeyRange(nil), explicitRanges...)
```

Related: `hasExplicitRanges` is true whenever the `ranges` key is present, including `"ranges": []`. `getKeyRanges([])` returns the whole key space, so a request that lists an already-configured store together with a new one silently resets the existing store's ranges. Omitting `ranges` preserves them (`TestEvictLeaderUpdateConfigBatchPreservesExistingRanges`); an empty array does not — also confirmed with a unit test.

Treat an empty `ranges` array as omitted (or reject it), and only overwrite when the client actually supplied range pairs.

Both cases pass against the current code as written (i.e. they document the current behavior, not the desired one):

```go
func TestEvictLeaderUpdateConfigBatchSharesExplicitRangesSlice(t *testing.T) {
	re := require.New(t)
	cancel, _, tc, oc := prepareSchedulersTest()
	defer cancel()

	tc.AddLeaderStore(1, 0)
	tc.AddLeaderStore(2, 0)

	sl, err := CreateScheduler(types.EvictLeaderScheduler, oc, storage.NewStorageWithMemoryBackend(),
		ConfigSliceDecoder(types.EvictLeaderScheduler, []string{"1"}), func(string) error { return nil })
	re.NoError(err)
	conf := sl.(*evictLeaderScheduler).conf

	body, err := json.Marshal(map[string]any{"store_ids": []int{1, 2}, "ranges": []string{"c", "d"}})
	re.NoError(err)
	req := httptest.NewRequest(http.MethodPost, "/config", bytes.NewReader(body))
	resp := httptest.NewRecorder()
	sl.ServeHTTP(resp, req)
	re.Equal(http.StatusOK, resp.Code)

	expected := []keyutil.KeyRange{keyutil.NewKeyRange("c", "d")}
	re.Equal(expected, conf.StoreIDWithRanges[1])
	re.Equal(expected, conf.StoreIDWithRanges[2])

	// applyStoreIDs assigns the same explicitRanges slice to every store.
	// Mutating one store's range therefore also mutates the other.
	conf.StoreIDWithRanges[1][0].StartKey = []byte("zzz")
	re.Equal([]byte("zzz"), conf.StoreIDWithRanges[2][0].StartKey,
		"store 1 and store 2 share the same ranges backing array")
}

func TestEvictLeaderUpdateConfigEmptyRangesOverwritesExisting(t *testing.T) {
	re := require.New(t)
	cancel, _, tc, oc := prepareSchedulersTest()
	defer cancel()

	tc.AddLeaderStore(1, 0)
	tc.AddLeaderStore(2, 0)

	sl, err := CreateScheduler(types.EvictLeaderScheduler, oc, storage.NewStorageWithMemoryBackend(),
		ConfigSliceDecoder(types.EvictLeaderScheduler, []string{"1", "a", "b"}), func(string) error { return nil })
	re.NoError(err)
	conf := sl.(*evictLeaderScheduler).conf
	customRanges := conf.StoreIDWithRanges[1]
	re.Equal([]keyutil.KeyRange{keyutil.NewKeyRange("a", "b")}, customRanges)

	// An empty "ranges" array is present in the JSON, so hasRanges is true.
	// getKeyRanges([]) returns the whole key space and overwrites store 1.
	body, err := json.Marshal(map[string]any{"store_ids": []int{1, 2}, "ranges": []string{}})
	re.NoError(err)
	req := httptest.NewRequest(http.MethodPost, "/config", bytes.NewReader(body))
	resp := httptest.NewRecorder()
	sl.ServeHTTP(resp, req)
	re.Equal(http.StatusOK, resp.Code)

	wholeKeySpace := []keyutil.KeyRange{keyutil.NewKeyRange("", "")}
	re.Equal(wholeKeySpace, conf.StoreIDWithRanges[1],
		"empty ranges array should not reset an existing store's custom ranges")
	re.Equal(wholeKeySpace, conf.StoreIDWithRanges[2])
}
```

File: tools/pd-ctl/pdctl/command/scheduler_command.go [hunk 10]
```diff
@@ -688,7 +697,11 @@ func newConfigShuffleRegionCommand() *cobra.Command {
 }
 
 func addStoreToSchedulerConfig(cmd *cobra.Command, schedulerName string, args []string) {
-	if len(args) != 1 {
+	// Only evict-leader-scheduler accepts more than one store id; this
+	// function is also reachable directly from other schedulers' own
+	// "config <name> add-store" subcommand (e.g. grant-leader-scheduler),
+	// which must keep requiring exactly one.
+	if len(args) < 1 || (schedulerName != evictLeaderSchedulerName && len(args) != 1) {
```
Comment (lhy1024) [inline tools/pd-ctl/pdctl/command/scheduler_command.go:704]: Nit: this now accepts multiple store IDs for `evict-leader-scheduler`, but `newConfigEvictLeaderCommand`'s `add-store` `Use` is still `add-store <store-id>`. `scheduler add evict-leader-scheduler` was updated to `<store_id> [store_id...]`; the config subcommand help should match. A `scheduler config evict-leader-scheduler add-store 3 4` test would also be useful — the existing pd-ctl coverage only exercises `scheduler add ... 3 4`.

The help-text mismatch is not worth a unit test by itself, but the missing multi-store `add-store` path can be covered next to `checkEvictLeaderSchedulerMultiStore`, e.g.

```go
output, err := tests.ExecuteCommand(cmd, []string{"-u", pdAddr, "scheduler", "config", "evict-leader-scheduler", "add-store", "3", "4"}...)
re.NoError(err)
re.Contains(string(output), "Success!")
for _, storeID := range []string{"3", "4"} {
	testutil.Eventually(re, func() bool {
		output, err = tests.ExecuteCommand(cmd, []string{"-u", pdAddr, "store", storeID}...)
		re.NoError(err)
		storeInfo := new(response.StoreInfo)
		re.NoError(json.Unmarshal(output, &storeInfo))
		return storeInfo.Status.PauseLeaderTransferIn && !storeInfo.Status.PauseLeaderTransferOut
	})
}
```

**Review by lhy1024 — COMMENTED** [review]
Two small follow-ups on the unified config-update path (ranges handling) and the pd-ctl `add-store` help text. The batch apply/rollback itself looks good.

File: pkg/schedule/schedulers/evict_leader.go (hunk 9 above)
Comment (bufferflies) [reply pkg/schedule/schedulers/evict_leader.go]: Good catch, fixed in `00e2c11f9`: `applyStoreIDs` now copies `explicitRanges` into a fresh slice per store (`append([]keyutil.KeyRange(nil), explicitRanges...)`) instead of assigning the same slice to every store in the batch, so mutating one store's range can no longer affect another. Added `TestEvictLeaderUpdateConfigBatchCopiesExplicitRangesPerStore`, adapted from your repro, which fails without the fix.

Also fixed the empty-`ranges`-array case: `updateConfig` now treats `"ranges": []` the same as an omitted field (only calls `getKeyRanges` when the decoded slice is non-empty), so it no longer resets an already-configured store to the whole key space when batched with a new store. Added `TestEvictLeaderUpdateConfigEmptyRangesPreservesExisting`, adapted from your repro, which fails without the fix.

File: tools/pd-ctl/pdctl/command/scheduler_command.go (hunk 10 above)
Comment (bufferflies) [reply tools/pd-ctl/pdctl/command/scheduler_command.go:704]: Good catch, fixed in `00e2c11f9`: `newConfigEvictLeaderCommand`'s `add-store` `Use` string now reads `add-store <store_id> [store_id...]`, matching `scheduler add evict-leader-scheduler <store_id> [store_id...]`. Also added a `scheduler config evict-leader-scheduler add-store 3 4` test next to `checkEvictLeaderSchedulerMultiStore`, adapted from your example, verifying the multi-store path through the config subcommand.

File: server/api/scheduler.go [hunk 11]
```diff
@@ -207,6 +209,62 @@ func (h *schedulerHandler) CreateScheduler(w http.ResponseWriter, r *http.Reques
...
+					// "create with N stores" request is atomic: either every
+					// requested store ends up evicted, or none of them do.
+					if rmErr := h.RemoveScheduler(name); rmErr != nil && !errors.ErrorEqual(rmErr, errs.ErrSchedulerNotFound.FastGenByArgs()) {
+						// The scheduler may still be active with only the
+						// first store evicted; surface both failures instead
+						// of silently reporting just the original one.
+						log.Error("failed to roll back partially created evict-leader-scheduler",
+							zap.String("scheduler-name", name), errs.ZapError(rmErr))
+						h.r.JSON(w, http.StatusInternalServerError, fmt.Sprintf(
+							"failed to add stores %v: %s; additionally failed to roll back the partially created scheduler: %s (it may still be running with only store %d evicted, please check and remove it manually if needed)",
```
Comment (rleungx) [inline server/api/scheduler.go]: **P2** — On the double-failure rollback path (`justCreated` + `RemoveScheduler` also failing), the scheduler may stay registered with store `storeIDs[0]` still evicted in memory. The "please check and remove it manually if needed" wording points at `RemoveScheduler` — the very call that just failed — so the suggested manual remediation re-invokes the broken operation. Suggest naming the concrete `rmErr` reason and pointing at the actual cleanup path (repair the persisted scheduler config), not a plain re-invoke. Wording-only fix; behavior unchanged.

File: server/api/scheduler.go [hunk 12]
```diff
@@ -207,6 +209,62 @@ func (h *schedulerHandler) CreateScheduler(w http.ResponseWriter, r *http.Reques
...
+							toUpdate, err.Error(), rmErr.Error(), uint64(storeIDs[0])))
+						return
+					}
+				}
+				h.r.JSON(w, http.StatusInternalServerError, err.Error())
```
Comment (rleungx) [inline server/api/scheduler.go:259]: **P3** — `RedirectSchedulerUpdateBatch` failure covers both "a batch store id doesn't exist" (a client error, since `applyStoreIDs` fails in `PauseLeaderTransfer`) and an actual `save()` persistence failure. This returns 500 for both. The missing-store case is arguably a 400. Minor status-code precision; the body still explains the cause, so non-blocking.

**Review by rleungx — APPROVED** [review]
The rest LGTM

File: pkg/schedule/schedulers/evict_leader.go (hunk 9 above)
Comment (lhy1024) [reply pkg/schedule/schedulers/evict_leader.go]: Confirmed on `00e2c11f9`. Per-store copy of `explicitRanges` and treating `"ranges": []` as omitted both look correct. I re-ran `TestEvictLeaderUpdateConfigBatchCopiesExplicitRangesPerStore` and `TestEvictLeaderUpdateConfigEmptyRangesPreservesExisting`; both pass.

File: tools/pd-ctl/pdctl/command/scheduler_command.go (hunk 10 above)
Comment (lhy1024) [reply tools/pd-ctl/pdctl/command/scheduler_command.go:704]: Confirmed on `00e2c11f9`. The `add-store` usage now matches `scheduler add`, and the extra `scheduler config evict-leader-scheduler add-store 3 4` coverage is what I was looking for. Thanks.

**Review by lhy1024 — APPROVED** [review]
The previous comments on per-store range copies, empty `ranges` handling, and the pd-ctl `add-store` usage are addressed in `00e2c11f9`. LGTM.

File: server/api/scheduler.go (hunk 11 above)
Comment (bufferflies) [reply server/api/scheduler.go]: Good catch, fixed in `b19412722`: the message now names the concrete rollback error explicitly ("resolve the rollback error above") and points at retrying the same removal afterward, instead of phrasing it as if a separate "manual" cleanup path existed — the only actual cleanup path is `RemoveScheduler` itself, so the wording now says that plainly rather than implying otherwise.

File: tools/pd-ctl/tests/scheduler/scheduler_test.go [hunk 13]
```diff
@@ -922,6 +922,80 @@ func (suite *schedulerTestSuite) checkEvictLeaderScheduler(cluster *pdTests.Test
...
+	})
+	for _, storeID := range []string{"3", "4"} {
+		testutil.Eventually(re, func() bool {
+			output, err = tests.ExecuteCommand(cmd, []string{"-u", pdAddr, "store", storeID}...)
+			re.NoError(err)
+			storeInfo := new(response.StoreInfo)
+			re.NoError(json.Unmarshal(output, &storeInfo))
+			return storeInfo.Status.PauseLeaderTransferIn && !storeInfo.Status.PauseLeaderTransferOut
+		})
+	}
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply tools/pd-ctl/tests/scheduler/scheduler_test.go]: Good catch, fixed in `b19412722`: the pre-removal pause-state loop now also checks store 1, so a hypothetical add-store bug that replaces the store set instead of appending to it would fail this test.

File: server/api/scheduler.go (hunk 12 above)
Comment (bufferflies) [reply server/api/scheduler.go:259]: Confirmed, but not changing this in this PR: `RedirectSchedulerUpdateBatch`'s 500-for-everything matches the pre-existing single-store `RedirectSchedulerUpdate` path (`server/api/scheduler.go:199`), which has the identical imprecision — this PR's batch path just inherited it. Distinguishing a client error (missing store) from a real persistence failure would need the original HTTP status code to survive `pkg/utils/apiutil`'s `checkResponse`/`PostJSONIgnoreResp`, which currently discard it; that's shared plumbing behind both paths, so fixing it here would mean touching both, not just the new batch code. Tracked in the PR description's Known limitations. Happy to open a follow-up issue if you'd like it tracked separately — let me know.
